### PR TITLE
numa: account for incorrect core number on topology.insert

### PR DESCRIPTION
**Pull request by pkazmierczak**
Unsupported environments like containers or guests OSs inside LXD can incorrectly number of available cores thus leading to numalib having trouble detecting cores and panicking. This code adds tests for linux sysfs detection methods and fixes the panic from #19372.

### CODE REVIEW

File: client/lib/numalib/detect_linux_test.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	"os"
+	"testing"
+
+	"github.com/hashicorp/nomad/client/lib/idset"
+	"github.com/hashicorp/nomad/client/lib/numalib/hw"
+	"github.com/shoenig/test/must"
+)
+
+// badSysData are example values from sysfs on unsupported platforms, e.g.,
+// containers, virtualization guests
```
Comment (tgross) [inline client/lib/numalib/detect_linux_test.go:18]: There's a good bit here, so it'd be helpful if there were a comment alongside the lines that are different indicating what's unexpected.

File: client/lib/numalib/detect_linux.go [hunk 2]
```diff
@@ -105,46 +109,50 @@ func (*Sysfs) discoverCores(st *Topology) {
...
 			if err != nil {
 				return err
 			}
 
 			cores := idset.Parse[hw.CoreID](string(s))
 			_ = cores.ForEach(func(core hw.CoreID) error {
+				// if we get incorrect core number, stop
+				if int(core) >= len(st.Cores) {
+					return nil
+				}
```
Comment (tgross) [inline client/lib/numalib/detect_linux.go:130]: This should stop the panic. But does it leave us in an inconsistent state elsewhere?

File: client/lib/numalib/detect_linux_test.go (hunk 1 above)
Comment (pkazmierczak) [reply client/lib/numalib/detect_linux_test.go:18]: I added extra comments

**Review by tgross — APPROVED** [review]
LGTM

**Pull request merged**